plugins/flashrom/fu-flashrom-lspcon-i2c-spi-device.c

```diff
@@ -0,0 +1,315 @@
+{
+	FuFlashromDevice *parent = FU_FLASHROM_DEVICE (device);
+	FuFlashromLspconI2cSpiDevice *self = FU_FLASHROM_LSPCON_I2C_SPI_DEVICE (device);
+	const gchar *path = NULL;
+	const gchar *hw_id = NULL;
+	g_autofree gchar *vid = NULL;
+	g_autofree gchar *pid = NULL;
+	g_autofree gchar *guid = NULL;
+	g_autofree gchar *contents = NULL;
+	struct flashrom_flashctx *flashctx = NULL;
+	gsize flash_size;
+	uint32_t addr;
+
+	if (g_strcmp0 (fu_flashrom_device_get_programmer_name (parent),
+		       "lspcon_i2c_spi") != 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "invalid programmer");
+		return FALSE;
+	}
+
+	path = fu_udev_device_get_sysfs_path (FU_UDEV_DEVICE (device));
+	self->bus_number = fu_flashrom_lspcon_i2c_spi_device_get_bus_number_from_path (path, error);
+	if (self->bus_number == -1) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "%s failed to get bus number for " \
+			     "device under %s.",
+			     __func__,
+			     path);
+		return FALSE;
+	}
+
+	hw_id = fu_udev_device_get_sysfs_attr (FU_UDEV_DEVICE (device), "name", error);
+	vid = g_strndup (hw_id, HID_LENGTH / 2);
+	pid = g_strndup (&hw_id[HID_LENGTH / 2], HID_LENGTH / 2);
+	fu_device_add_vendor_id (device, g_strdup_printf ("I2C:%s", vid));
+
+	guid = fwupd_guid_hash_string (g_strdup_printf (DEVICE_GUID_FORMAT, vid, pid));
+	if (guid == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "%s failed to generate guid.",
+			     __func__);
+		return FALSE;
+	}
+	fu_device_add_guid (device, guid);
+
+	/* FuFlashromLspconI2cSPiDevice->prepare */
+	if (!fu_flashrom_lspcon_i2c_spi_device_prepare (device,
+							FWUPD_INSTALL_FLAG_NONE,
+							error))
+		return FALSE;
+
+	/* set up flashrom layout */
+	flashctx = fu_flashrom_device_get_flashctx (parent);
+	flashrom_layout_set (flashctx, &layout);
+
+	/* read the current flash contents */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_READ);
+	flash_size = fu_flashrom_device_get_flash_size (parent);
+	contents = g_malloc0 (flash_size);
+	if (flashrom_image_read (flashctx, contents, flash_size)) {
+		g_set_error_literal (error, FWUPD_ERROR, FWUPD_ERROR_READ,
+				     "failed to read flash contents");
+		return FALSE;
+	}
+
+	/* grab the active partition */
+	self->active_partition = contents[entries[0].start];
+
+	if (self->active_partition != 1 && self->active_partition != 2) {
+		g_set_error_literal (error, FWUPD_ERROR, FWUPD_ERROR_READ,
+				     "failed to read active partition");
+		return FALSE;
+	}
+
+	/* find the current version for the active partition */
+	addr = entries[self->active_partition].start + VERSION_ADDR_OFFSET;
+	fu_device_set_version (device, g_strdup_printf ("%d.%d", contents[addr],
+							contents[addr + 2]));
+
+	/* FuFlashromDevice->cleanup */
+	if (!FU_DEVICE_CLASS
+	    (fu_flashrom_lspcon_i2c_spi_device_parent_class)->cleanup (device,
```

Comment:
cleanup? In setup()?

Reply:
I know it is ugly... I may want to create another function on flashrom_device that can be called here and on flashrom_device->cleanup... Basically we need to call the cleanup (i.e., shutting down programmer, etc) every time we finish accessing the flash because libflashrom does not support more than one context (i.e., a internal programmer context and a lspcon_i2c_spi programmer context) at a time...

Reply:
> I may want to create another function on flashrom_device

Is that not FuDevice->open() and ->close()?